### PR TITLE
[Do Not Merge] Initial merge of new C#/VB project system

### PR DESCRIPTION
**Pull request by davkean**
This PR will be merged after we've RI'd _future_ into the internal Visual Studio tree.

This is the initial merge of the new C#/VB project system based on the [Common Project System](https://github.com/microsoft/vsprojectsystem).

Make note, I welcome feedback across the code base, however, I'll be filing work items to track things such as naming, coding style, structure and bugs in the code and fixing those in future PRs, and not block the merge. However, I will fix issues that destabilize the product, block the build, etc.

This introduces 6 new binaries to the product:

__Host Agnostic Layer:__
* __Microsoft.VisualStudio.ProjectSystem.CSharp:__ _Components for editing C# projects agnostic of host._
* __Microsoft.VisualStudio.ProjectSystem.VisualBasic:__ _Components for editing Visual Basic projects agnostic of host._
* __Microsoft.VisualStudio.ProjectSystem.Managed:__ _Components that are shared between C# and VB projects agnostic of host_

__Visual Studio Layer:__
* __Microsoft.VisualStudio.ProjectSystem.CSharp.VS:__ _Components for editing C# projects within Visual Studio_
* __Microsoft.VisualStudio.ProjectSystem.VisualBasic.VS:__ _Components for editing Visual Basic projects within Visual Studio_
* __Microsoft.VisualStudio.ProjectSystem.Managed.VS:__ _Components that are shared between C# and VB projects within Visual Studio. We eventually expect this to be shared with other managed languages that move to CPS._

This project system _will not_ currently work with RTM bits of Visual Studio, as these components conflict with the current CPS bits, however, in the future you will be able to F5 from ProjectSystemSetup, which will deploy the project system locally and you can open a msbuildproj (temporary) to test out the project system. Tests also, unlike the rest of Roslyn (due to the scale), can be run directly from Team Explorer.

__Work identified below to be done in future PRs:__

- [ ] [Prevent projects from copying the same references, project references and nuget impls to the output](https://github.com/dotnet/roslyn/issues/6771)
- [ ] [Figure out whether project system should be using ConfigureAwait(true/false) on its await calls](https://github.com/dotnet/roslyn/issues/6770)
- [ ] [Code review feedback for merge of CPS](https://github.com/dotnet/roslyn/issues/6772)

### CODE REVIEW

**Comment by davkean** [comment]
tag @dotnet/roslyn-ide

**Comment by davkean** [comment]
Failures are due to packages being out of date, fixing via https://github.com/dotnet/roslyn/blob/master/docs/infrastructure/Updating%20NuGet%20ZIP%20file.

File: build/Targets/VSL.Imports.targets [hunk 1]
```diff
@@ -1,6 +1,20 @@
 <Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 
+  <!-- Turn off implict binary and project reference copying to prevent projects
```
Comment (jaredpar) [inline build/Targets/VSL.Imports.targets]: Is this change independent of CPS?  Essentially something we should be doing in master?

File: src/VisualStudio/ProjectSystem/DeployTestDependencies/DeployTestDependencies.csproj [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <SchemaVersion>2.0</SchemaVersion>
+    <ProjectGuid>{37BA82E6-9ABD-4ACA-AA26-2DFD39A359A5}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <RootNamespace>Microsoft.VisualStudio</RootNamespace>
+    <AssemblyName>ProjectSystem</AssemblyName>
+    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
+    <OutDir>$(OutDir)\ProjectSystem\Tests</OutDir>
```
Comment (jaredpar) [inline src/VisualStudio/ProjectSystem/DeployTestDependencies/DeployTestDependencies.csproj]: Why do we need a custom output directory?

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests.csproj [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectGuid>{F81A2B36-6763-42B8-8F4D-5199325D40C1}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>Microsoft.VisualStudio</RootNamespace>
+    <AssemblyName>Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests</AssemblyName>
+    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
+    <SignAssembly>true</SignAssembly>
+    <OutDir>$(OutDir)\ProjectSystem\Tests</OutDir>
```
Comment (jaredpar) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests.csproj]: Do all our unit test scripts look for unit test DLLS recursively?

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests/ProjectSystem/Designers/PropertiesFolderProjectTreeModifierTests.cs [hunk 4]
```diff
@@ -0,0 +1,448 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.VisualStudio.ProjectSystem.Designers.Imaging;
+using Microsoft.VisualStudio.Testing;
+using Xunit;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Designers
+{
+    [UnitTestTrait]
```
Comment (jaredpar) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests/ProjectSystem/Designers/PropertiesFolderProjectTreeModifierTests.cs]: Why is this needed?

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.VS.UnitTests/ProjectSystem/Designers/Imaging/CSharpProjectImageProviderTests.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        public void GetProjectImage_NullAsKey_ThrowsArgumentNull()
+        {
+            var provider = CreateInstance();
+
+            Assert.Throws<ArgumentNullException>("key", () => {
+
+                provider.GetProjectImage((string)null);
+            });
+        }
+
```
Comment (jaredpar) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.VS.UnitTests/ProjectSystem/Designers/Imaging/CSharpProjectImageProviderTests.cs]: Double blank line

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests.csproj (hunk 3 above)
Comment (davkean) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests.csproj]: Yes

File: build/Targets/VSL.Imports.targets (hunk 1 above)
Comment (davkean) [inline build/Targets/VSL.Imports.targets]: Plan to, wanted a staged approach. Will file a bug on this.

Comment (davkean) [inline build/Targets/VSL.Imports.targets]: Filed https://github.com/dotnet/roslyn/issues/6771.

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests/ProjectSystem/Designers/PropertiesFolderProjectTreeModifierTests.cs (hunk 4 above)
Comment (davkean) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.UnitTests/ProjectSystem/Designers/PropertiesFolderProjectTreeModifierTests.cs]: It groups the project system tests under single node in Test Explorer for easy run, I'll rename it to make sure it says something regarding project system (https://github.com/dotnet/roslyn/issues/6772)

File: src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.VS.UnitTests/ProjectSystem/Designers/Imaging/CSharpProjectImageProviderTests.cs (hunk 5 above)
Comment (davkean) [inline src/VisualStudio/ProjectSystem/Microsoft.VisualStudio.ProjectSystem.CSharp.VS.UnitTests/ProjectSystem/Designers/Imaging/CSharpProjectImageProviderTests.cs]: https://github.com/dotnet/roslyn/issues/6772

**Comment by davkean** [comment]
Linux box failure was due to Jenkins being taken down, the windows was due to out-of-date packages, which I've just merged in.

retest prtest/lin/dbg/unit32 please
retest prtest/win/dbg/unit32  please
retest prtest/win/dbg/unit64  please

**Comment by davkean** [comment]
For some reason the CPS properties task isn't running on build machine, waiting for my build log change to get merged in for me to diagnose and figure it out.

**Comment by davkean** [comment]
Close and then reopen to sync with the MSBuild log changes from future to help diagnose the windows failures.

**Comment by davkean** [comment]
@dotnet/roslyn-ide Heads up, I'm going to merge this in on Monday. 

Feel free to review it *after* it is in (it's probably easier then because you don't need to manually pull it down) - and add comments to: https://github.com/dotnet/roslyn/issues/6772

**Comment by Pilchie** [comment]
Some rough notes:
[ ] Typo in file name "Designers\Input\SingeNodeProjectCommandBase.cs"
[ ] Our normal convention is for internal abstract types to be prefixed with `Abstract` instead of suffixed with `Base`.
[ ] Doc comment on `ProjectRootImageProjectTreeModifier` is wrong.
[ ] Tests (in general) a lot of duplicated setup code.  Possible to extract into helpers so that tests focus on what is unique about that test?
[ ] CSharpProjectXmlProjectCapabilityProvider what does an [ImportingConstructor] do when there are no parameters?

**Comment by davkean** [comment]
Thanks @Pilchie, added those to the feedback work item here: https://github.com/dotnet/roslyn/issues/6772. I'll look at that on Monday.
